### PR TITLE
Negative log loss fn

### PR DESCRIPTION
**Pull request by danielvik**
## Description
Implementation of the loss function described by Lim et al. (2022) JCIM. Also described in this feature request: https://github.com/chemprop/chemprop/issues/612

## Example / Current workflow

Alternative loss function for model training. Detailed example is included in the training_NLL.ipynb.

## Questions
Uncertain about what is required in terms of documentation. Currently covered in the training_NLL.ipynb, but let me know if something more detailed should be provided, and where.

## Checklist
- [ ] linted with flake8?
- [ x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by danielvik** [comment]
I know exactly what went wrong. Should be fixed now 👍

**Review by JacksonBurns — DISMISSED** [review]
Thanks for opening this PR! The loss function looks mostly correct, but we need to make some other decisions before merging. I'll describe here in addition to a couple small notes in the code itself.

We don't need to define a new FFN subclass just for this code. Instead, we should just use the `output_transform` with torch.nn.Softplus (especially since, as the demo notebook says, one should not use an output_transform). I was able to run the notebook as provided and with these changes:

```python
ffn = nn.predictors.RegressionFFN(
    criterion=nn.metrics.NLogProbEnrichment(
        n1 = total_counts_pos,
        n2 = total_counts_neg,
        method="sqrt", 
        zscale=1.0,
    ),
    output_transform=Softplus(),
)
```

So I suggest removing `RegressionSoftplusFNN` entirely.

Right now this new loss function will be automatically exposed through the CLI but won't work properly since it needs additional input arguments (`n1` and `n2`).
We should also update the CLI code to check if this loss function is being used and then:
 - calculate `n1` and `n2` to properly initialize the loss function
 - add softplus as the FFN's output_transform as well (even if we don't end up making this change specifically, the CLI needs to check for the use of this loss function and adjust the FFN anyway)

Some other small things:
 - we need to add this demo notebook to this to get it to render in the online documentation: https://github.com/chemprop/chemprop/blob/main/docs/source/tutorial/python/index.rst
 - please make mention of which demo notebook it is based on and provide a link to it within the new notebook

Please let me know if you have any questions, and again thanks for getting the ball rolling on this one.

**Comment by danielvik** [comment]
I believe I resolved all the suggestions. However, I am very much in doubt about how to update index.rst correctly. The demo notebook training_NLL.ipynb is based on training.ipynb. 

For the CLI code, I'll need to dig into it, as I havn't had a look at it yet. It'll be some time before I can do that.

**Comment by JacksonBurns** [comment]
I'll take a look at the CLI code! Thanks for the updates

File: examples/training_NLL.ipynb [hunk 1]
```diff
@@ -0,0 +1,1332 @@
+{
+    "cells": [
+        {
+            "cell_type": "markdown",
+            "metadata": {},
+            "source": [
+                "# Training with Negative Log Loss as Loss function. \n",
```
Comment (KnathanM) [inline examples/training_NLL.ipynb]: Here the loss function is called "Negative Log Loss". This could cause confusion as torch as `torch.nn.NLLLoss` which is for classification tasks, while the loss function in this PR is for regression.

Adding something about enrichment or Poisson distribution could help. The name of the loss function is more specific `NLogProbEnrichment`.
This comment also applies to the name of this notebook.

File: chemprop/nn/metrics.py [hunk 2]
```diff
@@ -580,3 +582,77 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+    """
+
+    def __init__(
+        self,
+        task_weights: ArrayLike = 1.0,
+        n1: int = 1,
+        n2: int = 1,
+        method: str = "sqrt",
+        zscale: float = 1.0,
+        zinterval: int = 5,
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/nn/metrics.py]: Fixed

File: tests/unit/nn/test_loss_functions.py [hunk 3]
```diff
@@ -501,4 +502,48 @@ def test_Wasserstein(
     torch.testing.assert_close(loss, expected_loss)
 
 
+@pytest.mark.parametrize(
+    "preds,targets,mask,weights,task_weights,n1,n2,expected_loss",
+    [
+        (
+            torch.tensor([[0, 2], [0, 1]], dtype=torch.float),  # preds
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply tests/unit/nn/test_loss_functions.py]: I believe it is `2x2` because of the positive and negative counts being in separate columns

**Comment by JacksonBurns** [comment]
@KnathanM I think I've addressed everything, let me know how it looks.

File: examples/training_NLogProbEnrichment.ipynb [hunk 4]
```diff
@@ -0,0 +1,1074 @@
...
+            "execution_count": 14,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "criterion = nn.metrics.NLogProbEnrichment(\n",
+                "    n1=total_counts_pos,\n",
+                "    n2=total_counts_neg,\n",
+                "    method=\"sqrt\", \n",
+                "    zscale=1.0,\n",
+                "    zinterval=5,\n",
```
Comment (KnathanM) [inline examples/training_NLogProbEnrichment.ipynb]: It really doesn't matter, but now the type hint for zinterval says it should be a float, but here it is now an int. Again, not really important.

**Review by KnathanM — COMMENTED** [review]
Thanks @JacksonBurns for the changes and @danielvik for starting it. The code looks ready to me. Just one or two comments on the tests.

File: tests/unit/nn/test_loss_functions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply tests/unit/nn/test_loss_functions.py]: Done in 96b0414

File: tests/unit/nn/test_loss_functions.py [hunk 5]
```diff
@@ -501,4 +502,45 @@ def test_Wasserstein(
...
+            torch.zeros([2, 2], dtype=torch.float),  # targets
+            torch.ones([2, 2], dtype=torch.bool),  # mask
+            torch.ones([2]),  # weights
+            torch.tensor(10, dtype=torch.int),  # n1
+            torch.tensor(1000, dtype=torch.int),  # n2
+            torch.tensor(2.7664, dtype=torch.float),  # exptected_loss
+        ),
+    ],
+)
+def test_NLL(preds, targets, mask, weights, n1, n2, expected_loss):
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply tests/unit/nn/test_loss_functions.py]: I've added the different `method` options in 6618587 - we can confirm that `score` raises an error, and then fix it.

**Comment by JacksonBurns** [comment]
With the merger of #1306 we should now be able to get this in!

File: chemprop/nn/metrics.py [hunk 6]
```diff
@@ -580,3 +583,77 @@ def _calc_unreduced_loss(self, preds: Tensor, targets: Tensor, mask: Tensor, *ar
...
+        if method == "score":
+            zstat = (k1 - k2 * R_d) / torch.sqrt((k1 + k2) * R_d)
+        elif method == "wald":
+            zstat = (k1 - k2 * R_d) / torch.sqrt(k1 + k2 * R_d**2)
+        elif method == "sqrt":
+            zstat = 2 * (torch.sqrt(k1 + 3 / 8.0) - torch.sqrt((k2 + 3 / 8.0) * R_d))
+            zstat = zstat / torch.sqrt(1 + R_d)
+        else:
+            raise NotImplementedError
+        return zstat
```
Comment (JacksonBurns) [inline chemprop/nn/metrics.py]:
```suggestion
        if torch.isnan(zstat).any():
            raise ValueError(f"Got nan loss with method {self.method}, ensure row-wise sum of counts are nonzero or use method='sqrt'")
        return zstat
```

**Comment by KnathanM** [comment]
I should also note, the expected test values were calculated using the code, so this test is more about "Does this function give the same value as when it was code" instead of "Does this function give the right theoretical value"

Ready for review @JacksonBurns
